README.md
# ECMA5 Multi-Purpose Javascript Toolkit
Documentation is at: jstoolkit.org

## Features
* Emulates ECMA5 Object features in legacy browsers. http://ejohn.org/blog/ecmascript-5-objects-and-properties/
* Over 86 native object prototypes extensions.
* Fully portable and tested on multiple platforms.

## Design Philosophy
* Written in pure optimised javascript.
* Use JST methods within JST wherever possible.
* Lightweight, fast and small. Suitable for mobile devices.

## Avoided
* Overlapping frameworks like JQuery, Prototype etc.
* Full i18n / l10n compatibility. Although 90% of JST methods are locale-independent, JST is english-only.
* Monolithic code-base and be-all-and-end-all integrated library.

## Installation and Getting Started

### Browser
If you plan to use JSToolkit in the browser it is advised that you link `jst.min.js` in a script tag in your header. You can find `jst.min.js` in the `dist/` folder of this repository or from the website jstoolkit.org and click the download button.

### Node.JS / NPM
Using JSToolkit in Node.JS is incredibly easy. Simply install the package using a simple command:

    npm install toolkit -g

And to setup the integrated methods simply put `require('toolkit')` anywhere within your script.

#### Node.JS REPL
Because of the way Node.JS's REPL works, you cannot use toolkit as you would normally so we have provided a special integrated repl that works with JSToolkit.

    $ node
    > require('toolkit').repl();
    toolkit> Array.range(1,9,3);
    [ 1, 4, 7 ]
    
## Goals & Features
The JS Toolkit has very specific goals - as to not re-invent the wheel. They go as follows:

* Complete Unit-Test coverage.
* Designed entirely for ECMA5 applications - Node.JS/V8 specifically.
* Performance is key and is achieved through elegance and simplicity.
* ECMA5 introduced a number of new Object definition descriptors which must be respected.
* Eases development by extending type prototypes safely. Object.prototype is left untouched.
* Use accelerated native functions and fully embrace the new ECMA5 specification.

## Contributing
Please follow the following steps to contributing.

1. Check whether an issue containing your fix/feature is already open, if not create one.
2. Fork the JSToolkit repository and create a branch for fix. Like `feature/Array#foobar` or `bugfix/Array#diff`
3. Make all commits relating to your issue to that branch and once ready put it into a pull request referencing the issue.
4. Be sure to include full unit-tests for your change and ensure they pass! We use QUnit as our testing framework.

## License
(The MIT License)

Copyright (c) 2011 Oliver Morgan <oliver.morgan@kohark.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

HISTORY.md
1.5.4 / 2011-09-30
==================

  * Fixed multiple bugs on a number of un-tested functions.
  * Loads of doc typos fixed and examples changed to work with the online repl.

1.5.3 / 2011-09-30
==================

  * Major fixes to the online REPL
  * Added loads more tests to test edge-cases.
  * Array#Swap => Array#Swap$ - Was considered a bug in the API which needed changing
  * Added new methods Array#Swap, Date#fuzzyDiff, Number#ordinal, RegExp.escape, Number#pow, Object.follow
  * Number#ceil and Number#floor now accept precision parameters.
  * Deprecate Number#odd and Number#even these will be re-named to isOdd and isEven in 1.6.0  

1.5.2 / 2011-09-26
==================

  * Urgent fix for Node.JS platform.

1.5.1 / 2011-09-26
==================

  * New AutoDoc generator and website.
  * Added: String#repeat.
  * Number of small bug fixes.

1.5.0 / 2011-09-25
==================

  * Migrated testing framework to QUnit.
  * Added all missing ECMA5 methods from ECMA3.
  * Added: Array#concat$, String#chunk
  * Added advanced descriptor processing utilities for legacy browsers.
  * Fixed to pass tests on Safari 5.1, Chrome 14 and IE 5.5, 6, 7 and 8

1.4.0 / 2011-09-18 
==================

  * Migrated testing framework to expresso.
  * Preparing for browser migration and updated a lot of the docs.
  * Added: New useful Array#plug key array parameter.
  * Added: Array shortcut utilities (avoids having to use Func.apply)
  * Added: Object.enum/Object.enum$ for working with enumerable hashes.
  * Bugfix: Multiple bug fixes to tests and methods.

1.3.0 / 2011-09-10 
==================

  * Brand new documenter to better generate docs.
  * A lot of random bug fixes and improvements to docs overall.
  * Added: 15 new methods including: Object.size, Array#invoke, Array#pluck, Array#grep, Array#sortBy, Array#fetch, Number#radix, Number#bin, Number#oct, Number#dec, Number#hexl, Number#hex, Number#abbr, String#pad and String#sprintf.
  * Added: 14 Self-modification functions including: Object.Map$, Object.filter$, Object.clean$, Array#map$, Array#Filter$, Array#clean$, Array#shuffle$, Array#chunk$, Array#flatten$, Array#invoke$, Array#pluck$, Array#grep, Array#sort$ and Array#sortBy$.
  * Bugfix: Object.reduce not working correctly.

1.2.0 / 2011-09-07 
==================

  * A number of improvements on the docs with a number of examples / descriptions updated.
  * Api: First() and Last() in Array.prototype now return just the value if given no parameter.
  * Added: String.prototype methods.
  * Added: Function.prototype methods.
  * Added: Object hash id. Useful for caching methods.
  * Bugfix: Sum() and Product() in Array.prototype didn't cast values.
  * Bugfix: Infamous error in the calculation of odd numbers.

1.1.0 / 2011-09-01 
==================

  * New documenter added to automatically generate documents
  * Added: Number.prototype methods added
  * Added: Object.walk method added
  * Bugfix: Object.clone wasn't working
  * Bugfix: Callbacks weren't respecting scopes
  * Bugfix: Object.map & filter recreated object incorrectly
  * Bugfix: Array.range was declared as enumerable

1.0.0 / 2011-08-27 
==================

  * Initial Release
  * Unit Tests Complete
  * Math, Array and Object extensions pass all tests and are to spec design.

js/inspect.js
/**
  @preserve
  Copyright Joyent, Inc. and other Node contributors.

  Permission is hereby granted, free of charge, to any person obtaining a
  copy of this software and associated documentation files (the
  "Software"), to deal in the Software without restriction, including
  without limitation the rights to use, copy, modify, merge, publish,
  distribute, sublicense, and/or sell copies of the Software, and to permit
  persons to whom the Software is furnished to do so, subject to the
  following conditions:

  The above copyright notice and this permission notice shall be included
  in all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
  OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
  NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
  DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
  OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
  USE OR OTHER DEALINGS IN THE SOFTWARE.

  Original at: https://github.com/joyent/node/blob/master/lib/util.js
*/

(function(){

// The maximum length of a line in the stylized output.
var MAX_COLUMNS = 80;

/**
 * Echos the value of a value. Tries to print the value out in the best way
 * possible given the different types.
 *
 * @param {Object} obj The object to print out.
 * @param {Boolean} showHidden Flag that shows hidden (not enumerable)
 *    properties of objects.
 * @param {Number} depth Depth in which to descend in object. Default is 2.
 * @param {Boolean} colors Flag to turn on ANSI escape codes to color the
 *    output. Default is false (no coloring).
 */
$.inspect = function(obj, showHidden, depth, colors) {
  var seen = [];

  var stylize = function(str, styleType) {
    // http://en.wikipedia.org/wiki/ANSI_escape_code#graphics
    var styles =
        { 'bold' : [1, 22],
          'italic' : [3, 23],
          'underline' : [4, 24],
          'inverse' : [7, 27],
          'white' : [37, 39],
          'grey' : [90, 39],
          'black' : [30, 39],
          'blue' : [34, 39],
          'cyan' : [36, 39],
          'green' : [32, 39],
          'magenta' : [35, 39],
          'red' : [31, 39],
          'yellow' : [33, 39] };

    var style =
        { 'special': 'cyan',
          'number': 'blue',
          'boolean': 'yellow',
          'undefined': 'grey',
          'null': 'bold',
          'string': 'green',
          'date': 'magenta',
          // "name": intentionally not styling
          'regexp': 'red' }[styleType];

    if (style) {
      return '\033[' + styles[style][0] + 'm' + str +
             '\033[' + styles[style][1] + 'm';
    } else {
      return str;
    }
  };
  if (! colors) {
    stylize = function(str, styleType) { return str; };
  }

  function format(value, recurseTimes) {
    // Primitive types cannot have properties
    switch (typeof value) {
      case 'undefined':
        return stylize('undefined', 'undefined');

      case 'string':
        var simple = '\'' + JSON.stringify(value).replace(/^"|"$/g, '')
                                                 .replace(/'/g, "\\'")
                                                 .replace(/\\"/g, '"') + '\'';
        return stylize(simple, 'string');

      case 'number':
        return stylize('' + value, 'number');

      case 'boolean':
        return stylize('' + value, 'boolean');
    }
    // For some reason typeof null is "object", so special case here.
    if (value === null) {
      return stylize('null', 'null');
    }

    // Look up the keys of the object.
    var visible_keys = Object.keys(value);
    var keys = showHidden ? Object.getOwnPropertyNames(value) : visible_keys;

    // Functions without properties can be shortcutted.
    if (typeof value === 'function' && keys.length === 0) {
      if (isRegExp(value)) {
        return stylize('' + value, 'regexp');
      } else {
        var name = value.name ? ': ' + value.name : '';
        return stylize('[Function' + name + ']', 'special');
      }
    }

    // Dates without properties can be shortcutted
    if (isDate(value) && keys.length === 0) {
      return stylize(value.toUTCString(), 'date');
    }

    var base, type, braces;
    // Determine the object type
    if (isArray(value)) {
      type = 'Array';
      braces = ['[', ']'];
    } else {
      type = 'Object';
      braces = ['{', '}'];
    }

    // Make functions say that they are functions
    if (typeof value === 'function') {
      var n = value.name ? ': ' + value.name : '';
      base = (isRegExp(value)) ? ' ' + value : ' [Function' + n + ']';
    } else {
      base = '';
    }

    // Make dates with properties first say the date
    if (isDate(value)) {
      base = ' ' + value.toUTCString();
    }

    if (keys.length === 0) {
      return braces[0] + base + braces[1];
    }

    if (recurseTimes < 0) {
      if (isRegExp(value)) {
        return stylize('' + value, 'regexp');
      } else {
        return stylize('[Object]', 'special');
      }
    }

    seen.push(value);

    var output = keys.map(function(key) {
      var name, str;
      if (value.__lookupGetter__) {
        if (value.__lookupGetter__(key)) {
          if (value.__lookupSetter__(key)) {
            str = stylize('[Getter/Setter]', 'special');
          } else {
            str = stylize('[Getter]', 'special');
          }
        } else {
          if (value.__lookupSetter__(key)) {
            str = stylize('[Setter]', 'special');
          }
        }
      }
      if (visible_keys.indexOf(key) < 0) {
        name = '[' + key + ']';
      }
      if (!str) {
        if (seen.indexOf(value[key]) < 0) {
          if (recurseTimes === null) {
            str = format(value[key]);
          } else {
            str = format(value[key], recurseTimes - 1);
          }
          if (str.indexOf('\n') > -1) {
            if (isArray(value)) {
              str = str.split('\n').map(function(line) {
                return '  ' + line;
              }).join('\n').substr(2);
            } else {
              str = '\n' + str.split('\n').map(function(line) {
                return '   ' + line;
              }).join('\n');
            }
          }
        } else {
          str = stylize('[Circular]', 'special');
        }
      }
      if (typeof name === 'undefined') {
        if (type === 'Array' && key.match(/^\d+$/)) {
          return str;
        }
        name = JSON.stringify('' + key);
        if (name.match(/^"([a-zA-Z_][a-zA-Z_0-9]*)"$/)) {
          name = name.substr(1, name.length - 2);
          name = stylize(name, 'name');
        } else {
          name = name.replace(/'/g, "\\'")
                     .replace(/\\"/g, '"')
                     .replace(/(^"|"$)/g, "'");
          name = stylize(name, 'string');
        }
      }

      return name + ': ' + str;
    });

    seen.pop();

    var numLinesEst = 0;
    var length = output.reduce(function(prev, cur) {
      numLinesEst++;
      if (cur.indexOf('\n') >= 0) numLinesEst++;
      return prev + cur.length + 1;
    }, 0);

    if (length > MAX_COLUMNS) {
      output = braces[0] +
               (base === '' ? '' : base + '\n ') +
               ' ' +
               output.join(',\n  ') +
               ' ' +
               braces[1];

    } else {
      output = braces[0] + base + ' ' + output.join(', ') + ' ' + braces[1];
    }

    return output;
  }
  return format(obj, (typeof depth === 'undefined' ? 2 : depth));
};

function isArray(ar) {
  return ar instanceof Array ||
         Array.isArray(ar) ||
         (ar && ar !== Object.prototype && isArray(ar.__proto__));
}

function isRegExp(re) {
  var s = '' + re;
  return re instanceof RegExp || // easy case
         // duck-type for context-switching evalcx case
         typeof(re) === 'function' &&
         re.constructor.name === 'RegExp' &&
         re.compile &&
         re.test &&
         re.exec &&
         s.match(/^\/.*\/[gim]{0,3}$/);
}

function isDate(d) {
  return d instanceof Date;
}

var months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep',
              'Oct', 'Nov', 'Dec'];

// 26 Feb 16:19:34
function timestamp() {
  var d = new Date();
  var time = [pad(d.getHours()),
              pad(d.getMinutes()),
              pad(d.getSeconds())].join(':');
  return [d.getDate(), months[d.getMonth()], time].join(' ');
}

})();